src/main/java/io/reactivex/internal/operators/flowable/FlowableTimeout.java

```diff
@@ -48,299 +47,343 @@ public FlowableTimeout(
+                task.dispose();
+
+                actual.onError(t);
+
+                task.dispose();
+            } else {
+                RxJavaPlugins.onError(t);
             }
-            long idx = index + 1;
-            index = idx;
+        }
 
-            if (!arbiter.onNext(t, s)) {
-                return;
+        @Override
+        public void onComplete() {
+            if (index.getAndSet(Long.MAX_VALUE) != Long.MAX_VALUE) {
+                task.dispose();
+
+                actual.onComplete();
+
+                task.dispose();
             }
+        }
 
-            Disposable d = timeout.get();
-            if (d != null) {
-                d.dispose();
+        @Override
+        public void onTimeout(long idx) {
+            if (index.compareAndSet(idx, Long.MAX_VALUE)) {
+                SubscriptionHelper.cancel(upstream);
+
+                Publisher<? extends T> f = fallback;
+                fallback = null;
+
+                long c = consumed;
+                if (c != 0L) {
+                    produced(c);
+                }
+
+                f.subscribe(new FlowableTimeoutTimed.FallbackSubscriber<T>(actual, this));
             }
+        }
 
-            Publisher<V> p;
+        @Override
+        public void onTimeoutError(long idx, Throwable ex) {
+            if (index.compareAndSet(idx, Long.MAX_VALUE)) {
+                SubscriptionHelper.cancel(upstream);
 
-            try {
-                p = ObjectHelper.requireNonNull(itemTimeoutIndicator.apply(t), "The publisher returned is null");
-            } catch (Throwable e) {
-                Exceptions.throwIfFatal(e);
-                actual.onError(e);
-                return;
+                actual.onError(ex);
+            } else {
+                RxJavaPlugins.onError(ex);
             }
+        }
 
-            TimeoutInnerSubscriber<T, U, V> tis = new TimeoutInnerSubscriber<T, U, V>(this, idx);
+        @Override
+        public void cancel() {
+            super.cancel();
+            task.dispose();
+        }
+    }
+
+    static final class TimeoutConsumer extends AtomicReference<Subscription>
+    implements FlowableSubscriber<Object>, Disposable {
+
+        private static final long serialVersionUID = 8708641127342403073L;
 
-            if (timeout.compareAndSet(d, tis)) {
-                p.subscribe(tis);
+        final TimeoutSelectorSupport parent;
+
+        final long idx;
+
+        TimeoutConsumer(long idx, TimeoutSelectorSupport parent) {
+            this.idx = idx;
+            this.parent = parent;
+        }
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            if (SubscriptionHelper.setOnce(this, s)) {
+                s.request(Long.MAX_VALUE);
+            }
+        }
+
+        @Override
+        public void onNext(Object t) {
+            if (get() != SubscriptionHelper.CANCELLED) {
+                get().cancel();
```

Comment:
Extract `get()` result to local var here?

Reply:
Sure.